streaming/base/dataset.py

```diff
@@ -277,86 +276,80 @@ def __init__(self,
 
-        # Remove the lock that makes it unpickleable
-        del self._worker_barrier.lock
-
-        # Set up the epoch counter.
+        # Epoch counter.
         #
         # Note: we do not assume that the end of __iter__() will ever be reached, so we need to
         # increment the epoch counter at the start of __iter__() instead of at the end, so we need
         # to track what the next epoch is, not the current epoch.
-        next_epoch_shm = CreateSharedMemory(name=f'{self._shm_prefix}_next_epoch',
-                                            size=np.int64().nbytes)
-        self._next_epoch_shm = next_epoch_shm.shm
-        self._next_epoch_arr = np.ndarray(1, buffer=self._next_epoch_shm.buf, dtype=np.int64)
-        self._next_epoch_arr[0] = 0
+        self._next_epoch = SharedScalar(np.int64, f'{self._shm_prefix}_next_epoch')
+
+        # Cache filelock. Proects downloading and evicting shards.
+        self._cache_filelock_path = os.path.join(self._filelock_root, '_cache_filelock')
+        self._cache_filelock: FileLock
+
+        # Cache usage.
+        self._cache_usage = SharedScalar(np.int64, f'{self._shm_prefix}_cache_usage')
 
-        # Get the filelock filename that protects shard_states shared memory array.
-        self.shard_states_filename = os.path.join(self._filelock_root, '_shard_states_filelock')
+        # Shard states array. Tells if a shard is missing, downloading, or present (eviction
+        # happens under the lock).
+        self._shard_states = SharedArray(self.num_shards, np.uint8,
+                                         f'{self._shm_prefix}_shard_states')
 
-        # Create or attach shard_states array (tells if each shard is unknown, downloading, or
-        # downloaded).
-        shard_states_shm = CreateSharedMemory(name=f'{self._shm_prefix}_shard_states',
-                                              size=self.num_shards * np.uint8(0).nbytes)
-        self._shard_states = shard_states_shm.shm
+        # Time of last access per shard. This is used to decide which shard(s) to evict when we run
+        # out of space.
+        self._shard_access_times = SharedArray(self.num_shards, np.float32,
+                                               f'{self._shm_prefix}_shard_access_times')
+
+        # Initialize shared memory objects.
+        if world.is_local_leader:
+            # Set initial epoch (before any resumption).
+            self.next_epoch = 0
+
+            # Init each stream's local dir, discovering which shards are present.
+            are_shards_present = []
+            for stream in self.streams:
+                stream_shards = stream.get_shards(world)
+                are_shards_present += stream.init_local_dir(stream_shards)
+
+            # Use that to init shard states, shard access times, and cache usage.
+            self.cache_usage = 0
+            for shard_id, is_shard_present in enumerate(are_shards_present):
```

Comment:
is this shard id consistent?

Reply:
yep -- (list of streams in provided order) -> (list of shard entries in provided order in the index.json for each stream)